pkg/clusteragent/admission/mutate/common.go

```diff
@@ -0,0 +1,79 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubeapiserver
+
+package mutate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"gomodules.xyz/jsonpatch/v3"
+	admiv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type mutateFunc func(*corev1.Pod) error
+
+// mutate handles mutating pods and encoding and decoding admission
+// requests and responses for the public mutate functions
+func mutate(req *admiv1beta1.AdmissionRequest, m mutateFunc) (*admiv1beta1.AdmissionResponse, error) {
+	var pod corev1.Pod
+	if err := json.Unmarshal(req.Object.Raw, &pod); err != nil {
+		return nil, fmt.Errorf("failed to decode raw object: %v", err)
+	}
+
+	if err := m(&pod); err != nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode the mutated Pod object: %v", err)
+	}
+
+	patchOperation, err := jsonpatch.CreatePatch(req.Object.Raw, bytes)
```

Comment:
Maybe we should leave a TODO to directly generate the patch instead of encode/decode/diff -> encode patch.
The POD JSON are usually pretty big and, in busy clusters, it might use a significant amount of CPU (to be benchmarked).

Reply:
good point